src/OpenLoco/src/Vehicles/VehicleBogie.cpp

```diff
@@ -156,20 +161,109 @@ namespace OpenLoco::Vehicles
+            }
 
-        call(0x004AA97A, regs);
-        return regs.dl != 1;
-    }
+            const auto tile = World::TileManager::get(World::Pos2{ tileX, tileY });
+            for (auto& el : tile)
+            {
+                auto* elTrack = el.as<World::TrackElement>();
+                if (elTrack == nullptr)
+                {
+                    continue;
+                }
 
-    // 0x004AF16A
-    void VehicleBogie::carComponent_sub_4AF16A()
-    {
-        registers regs;
-        regs.esi = X86Pointer(this);
+                if (elTrack->baseZ() != tileBaseZ)
+                {
+                    continue;
+                }
+
+                if (elTrack->unkDirection() != trackAndDirection.track.cardinalDirection())
+                {
+                    continue;
+                }
+
+                if (elTrack->trackId() != trackAndDirection.track.id())
+                {
+                    continue;
+                }
 
-        call(0x004AF16A, regs);
+                for (auto i = 0; i < 4; ++i)
+                {
+                    if (elTrack->hasMod(i) && trackObj->mods[i] == vehObj->rackRailType)
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+        else if (mode == TransportMode::road)
+        {
+            if (trackType == 0xFFU)
+            {
+                return true;
+            }
+            auto* roadObj = ObjectManager::get<RoadObject>(trackType);
+            if (!roadObj->hasFlags(RoadObjectFlags::unk_05))
+            {
+                return true;
+            }
+
+            auto* vehObj = ObjectManager::get<VehicleObject>(objectId);
+            if (!vehObj->hasFlags(VehicleObjectFlags::rackRail))
+            {
+                return false;
+            }
+
+            const auto tile = World::TileManager::get(World::Pos2{ tileX, tileY });
+            for (auto& el : tile)
+            {
+                auto* elRoad = el.as<World::RoadElement>();
+                if (elRoad == nullptr)
+                {
+                    continue;
+                }
+
+                if (elRoad->baseZ() != tileBaseZ)
+                {
+                    continue;
+                }
+
+                if (elRoad->unkDirection() != trackAndDirection.road.cardinalDirection())
+                {
+                    continue;
+                }
+
+                if (elRoad->roadId() != trackAndDirection.road.id())
+                {
+                    continue;
+                }
+
+                for (auto i = 0; i < 2; ++i)
+                {
+                    if (elRoad->hasMod(i) && roadObj->mods[i] == vehObj->rackRailType)
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+        else
+        {
+            assert(false);
+            return true;
+        }
```

Comment:
Given how big this function is would be perhaps best to split it up, its essentially two functions.